server/src/main.ts

```diff
@@ -255,17 +258,50 @@ export class TextDocuments {
 		connection.onDidChangeTextDocument((event: DidChangeTextDocumentParams) => {
 			let td = event.textDocument;
 			let changes = event.contentChanges;
-			let last: TextDocumentContentChangeEvent | undefined = changes.length > 0 ? changes[changes.length - 1] : undefined;
-			if (last) {
-				let document = this._documents[td.uri];
-				if (document && isUpdateableDocument(document)) {
-					if (td.version === null || td.version === void 0) {
-						throw new Error(`Received document change event for ${td.uri} without valid version identifier`);
-					}
-					document.update(last, td.version);
+			if (changes.length === 0) {
+				return;
+			}
+
+			let document = this._documents[td.uri];
+			if (!document || !isUpdateableDocument(document)) {
+				return;
+			}
+
+			const { version } = td;
+			if (version == null || version === void 0) {
+				throw new Error(`Received document change event for ${td.uri} without valid version identifier`);
+			}
+
+			switch (this.syncKind) {
+				case TextDocumentSyncKind.Full:
+					let last = changes[changes.length - 1];
+					document.update(last, version);
 					this._onDidChangeContent.fire(Object.freeze({ document }));
-				}
+					break;
+				case TextDocumentSyncKind.Incremental:
+					let edits: TextEdit[] = changes.map(change => {
+						let { text, range } = change;
+						if (range === null || range === void 0) {
+							throw new Error(`Received document change has null range for ${td.uri}`);
+						}
+						return { newText: text, range };
+					});
+
+					let updatedDoc: UpdateableDocument = edits.reduce(
+						(workingTextDoc: UpdateableDocument, edit) => {
+							let newText = TextDocument.applyEdits(workingTextDoc, [edit]);
+							workingTextDoc.update({ text: newText }, version);
```

Comment:
Currently every `update` on the TextDocument will flush the `_lineOffset`

https://github.com/Microsoft/vscode-languageserver-node/blob/9de8809400b7c443bc25f211dfaddcb74675b2d7/types/src/main.ts#L2654

while `applyEdits` relies on `_lineOffset` https://github.com/Microsoft/vscode-languageserver-node/blob/9de8809400b7c443bc25f211dfaddcb74675b2d7/types/src/main.ts#L2500-L2501

It means for every content change we try to apply, it will go through the whole document, character by character, to find all line breaks. In real world scenarios it can be really costly.

Reply:
Yup, I definitely observed this limitation.

These black-box functions (`applyEdits` / `update`) are the only ways to currently update text documents.  As we discussed in #476, I can use your npm module in this PR and swap out the implementations for these core functions as well.

Apologies, when I originally wrote this up, I wasn't aware that you had already published the package to npm.

Reply:
No worries and you are doing a great job here. I'm just a bit surprised to see that `lineOffsets` were being calculated every time when we do full sync (the good part is we calculate only when reading line info).

Based on your proposal and this PR, I think we can see two major performance improvement

* We don't need to send full document from client to server anymore when the sync mode is incremental
* The line offset calculation is incremental when the sync mode is incremental, which is actually the major time spent in loading a file in VS Code.

Let me know if you need any help from me, I can push directly to your PR with piece tree related changes (if you are okay with that).

Reply:
If you've made some sort of local progress on adding your piece tree as a dependency and using it, then feel free to push it to this PR.  Otherwise, I was planning on adding it as a dependency + using it in `FullTextDocument` / `TextDocument` namespace today.

Feel free to let me know, and we can proceed from there.

Reply:
@Jma353 go for it, I don't have anything yet, look forward to your change.

Reply:
For sure - I'll update this today or tomorrow.